File: backend/app/schemas/job.py
"""Job DTO"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.core.enums import normalize_job_status


class JobResponse(BaseModel):
    id: UUID
    project_id: UUID
    floor_id: Optional[UUID] = None
    job_type: str
    status: str
    input_json: dict[str, Any] = Field(default_factory=dict)
    result_json: dict[str, Any] = Field(default_factory=dict)
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime

    @field_validator("status", mode="before")
    @classmethod
    def _normalize_status(cls, v: Any) -> Any:
        """내부 status(running/done/failed/queued/completed) → API 표기로 통일.

        floorplan/rf/legacy 잡이 제각각 쓰는 종료 상태를 API 경계에서
        pending/running/succeeded/failed 로 normalize. DB 값은 그대로 둠.
        """
        if isinstance(v, str):
            return normalize_job_status(v)
        return v

File: frontend/src/types/job.ts
import type { ISODateString, UUID } from './common';

// 백엔드 §15 Job DTO (app/schemas/job.py 의 JobResponse 와 매칭)
// 도면 분석은 비동기로 큐잉되어 이 모델로 상태 추적됨.

export type JobStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'done'
  | 'failed'
  | string;

/** 백엔드 응답의 상태값 중 "완료" 로 간주되는 것들 (명세는 succeeded, 실제 응답은 done). */
const TERMINAL_SUCCESS = new Set(['succeeded', 'done']);
const TERMINAL_FAILURE = new Set(['failed', 'error']);

export function isJobSucceeded(status: string | null | undefined): boolean {
  return !!status && TERMINAL_SUCCESS.has(status);
}

export function isJobFailed(status: string | null | undefined): boolean {
  return !!status && TERMINAL_FAILURE.has(status);
}

export function isJobTerminal(status: string | null | undefined): boolean {
  return isJobSucceeded(status) || isJobFailed(status);
}

export interface Job {
  id: UUID;
  project_id: UUID;
  floor_id: UUID | null;
  job_type: string;
  status: JobStatus;
  input_json: Record<string, unknown>;
  result_json: Record<string, unknown>;
  error_message: string | null;
  started_at: ISODateString | null;
  finished_at: ISODateString | null;
  created_at: ISODateString;
}

/** 도면 분석 완료(succeeded) 후 result_json 에서 추출되는 값들. */
export interface FloorplanJobResult {
  scene_draft_id?: UUID;
}
